packages/clay-table/stories/index.tsx

```diff
@@ -534,4 +534,71 @@ storiesOf('Components|ClayTable', module)
 				</ClayTable.Body>
 			</ClayTable>
 		</form>
-	));
+	))
+	.add('with data table', () => {
+		const headers = [
+			{
+				key: 'checked',
+				type: 'checkbox',
+			},
+			{
+				header: 'Image',
+				key: 'imageUrl',
+				type: 'image',
+			},
+			{
+				expanded: true,
+				header: 'Teams',
+				headingTitle: true,
+				key: 'teams',
+			},
+			{
+				header: 'Region',
+				key: 'region',
+			},
+		];
+		const rows = [
+			{
+				checked: false,
+				id: 'a',
+				imageUrl: thumbnail,
+				region: 'South America',
+				teams: 'White and Red',
+			},
+			{
+				checked: true,
+				id: 'b',
+				imageUrl: thumbnail,
+				region: 'Europe',
+				teams: 'White and Purple',
+			},
+		];
+
+		const Image: React.FunctionComponent<any> = ({value, ...props}) => (
+			<ClayTable.Cell {...props}>
+				<img alt="thumbnail" className="table-img" src={value} />
+			</ClayTable.Cell>
+		);
+
+		const Checkbox: React.FunctionComponent<any> = ({value, ...props}) => (
+			<ClayTable.Cell {...props}>
+				<ClayCheckboxWithState
+					aria-label="Select row"
+					checked={value}
+				/>
+			</ClayTable.Cell>
+		);
+
+		const cellRenderers = {
```

Comment:
How are `cellRenderers` meant to be selected? Looking at code, it looks like `type` should do the trick, but I don't see any such field in the sample model here...

Reply:
The field above image and checkbox: https://github.com/liferay/clay/pull/2721/files#diff-3bf0ae6ed7a94d91023cf99cb7dc1ff2R540-R548

Is that what you are saying? `type` is in the prop of `header`.

Reply:
Yes, I didn't see that... I was expecting to see the renderer in the cell... it makes sense that type is set per column... I wonder if we'll ever have the case where we want cells in different rows to have different renderers...

Reply:
Hmm, I'm not sure if we would get a case down, probably not.

Reply:
Probably not.

One thing that will probably be needed is a `rowRenderer` concept and probably also the need for `cellRenderer` to control more than just the content. 

It's pretty common to either want a different `tr` with specific things (to customize a table) and a similar thing can be said for cells, where one might want to add/remove classes to the `td`, for example.

It's possible that this can already be achieved through `otherProps`, but we should validate the proper behaviour

Reply:
Yeah, essentially people can only control the whole cell with cellRenderer. I also thought about this a bit and people might want to add more things to the `tr` elements. The idea is that dev can override the default markup just by passing a new "layout" to children, for example:

```jsx
<ClayDataTable
	cellRenderers={cellRenderers}
	headers={headers}
	rows={rows}
>
	{(headers, rows, getCell) => (
		<Table>
			<Table.Head>
				<tr className="class-name">
					{headers.map(({header, key, ...props}) => (
						<Table.Cell
							{...Omit(props, ['type'])}
							headingCell
							key={key}
						>
							{header}
						</Table.Cell>
					))}
				</tr>
			</Table.Head>
			<Table.Body>
				{rows.map(row => (
					<Table.Row key={row.id}>
						{row.cells.map(({type, ...props}) =>
							getCell(type, props)
						)}
					</Table.Row>
				))}
			</Table.Body>
		</Table>
	)}
</ClayDataTable>
```

I think this approach is very valid, DataTable will deal with normalizing the data and we can also do more with the data in the future and dev will have full control of all the content, it seems very minimal and also decreases the amount of API's that we would have to create it so that we could change anything in the elements.